### PR TITLE
Refactor GMP and Poly dependencies

### PR DESCRIPTION
**Pull request by nafur**
Refactors GMP and libpoly to also use external projects and be available within cmake as proper targets.

### CODE REVIEW

File: cmake/FindGMP.cmake [hunk 1]
```diff
@@ -9,17 +9,80 @@
...
+  function(GetVersionPart OUTPUT FILENAME DESC)
+    file(STRINGS ${FILENAME} RES REGEX "^#define __GNU_MP_${DESC}[ \\t]+.*")
+    string(REGEX MATCH "[0-9]+" RES "${RES}")
+    set(${OUTPUT}
+        "${RES}"
+        PARENT_SCOPE
+    )
+  endfunction()
+  function(GetVersion OUTPUT FILENAME)
+    getversionpart(MAJOR "${FILENAME}" "VERSION")
```
Comment (mpreiner) [inline cmake/FindGMP.cmake:33]: Either use `GetVersionPart` here or rename the function to `getversionpart` (just for consistency).

File: cmake/FindGMP.cmake [hunk 2]
```diff
@@ -9,17 +9,80 @@
...
+    getversionpart(PATCH "${FILENAME}" "VERSION_PATCHLEVEL")
+    set(${OUTPUT}
+        "${MAJOR}.${MINOR}.${PATCH}"
+        PARENT_SCOPE
+    )
+  endfunction()
+  if(EXISTS "${GMP_INCLUDE_DIR}/gmp-x86_64.h")
+    getversion(GMP_VERSION "${GMP_INCLUDE_DIR}/gmp-x86_64.h")
+  else()
+    getversion(GMP_VERSION "${GMP_INCLUDE_DIR}/gmp.h")
```
Comment (mpreiner) [inline cmake/FindGMP.cmake:44]: Ditto

File: cmake/FindGMP.cmake [hunk 3]
```diff
@@ -9,17 +9,80 @@
...
+    getversionpart(MAJOR "${FILENAME}" "VERSION")
+    getversionpart(MINOR "${FILENAME}" "VERSION_MINOR")
+    getversionpart(PATCH "${FILENAME}" "VERSION_PATCHLEVEL")
+    set(${OUTPUT}
+        "${MAJOR}.${MINOR}.${PATCH}"
+        PARENT_SCOPE
+    )
+  endfunction()
+  if(EXISTS "${GMP_INCLUDE_DIR}/gmp-x86_64.h")
+    getversion(GMP_VERSION "${GMP_INCLUDE_DIR}/gmp-x86_64.h")
```
Comment (mpreiner) [inline cmake/FindGMP.cmake:42]: What systems have the `gmp-x86_64.h` header? Maybe add a comment here why it is needed? And why is it not part of the `find_path(GMP_INCLUDE_DIR NAMES gmp.h gmpxx.h)` call?

File: cmake/FindPoly.cmake [hunk 4]
```diff
@@ -9,26 +9,120 @@
...
+  string(REGEX MATCH "[0-9.]+" Poly_VERSION "${Poly_VERSION}")
+
+  check_system_version("Poly")
+endif()
+
+if(NOT Poly_FOUND_SYSTEM)
+  include(ExternalProject)
+
+  set(Poly_VERSION "0.1.8")
+  set(Poly_VERSION "bae67639726f63ed508a30845108bfdac4a77546")
```
Comment (mpreiner) [inline cmake/FindPoly.cmake:41]: Do we want to have `0.1.8` or `bae67639726f63ed508a30845108bfdac4a77546`?

File: cmake/FindGMP.cmake (hunk 3 above)
Comment (nafur) [reply cmake/FindGMP.cmake:42]: I'm not completely sure (I copied this code from somewhere else, though I wrote it myself :-) ). IIRC, this was on CentOS...

Comment (mpreiner) [reply cmake/FindGMP.cmake:42]: Until now we never had an issue with FindGMP that we had until now. Should we remove this check?

**Review by mpreiner — APPROVED** [review]

**Pull request merged**